src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,845 @@
+        scores: torch.Tensor,
+        output_hidden_states: Optional[bool] = False,
+    ) -> torch.Tensor:
+        keypoints = keypoints.transpose(1, 2)
+        scores = scores.unsqueeze(1)
+        inputs = torch.cat([keypoints, scores], dim=1)
+        return self.encoder(inputs, output_hidden_states=output_hidden_states)
+
+
+class SuperGlueMultiHeadAttention(nn.Module):
+    def __init__(self, config: SuperGlueConfig):
+        super().__init__()
+        self.feature_dim = config.descriptor_dim
+        self.num_heads = config.num_heads
+        self.dim = self.feature_dim // self.num_heads
+        self.merge = nn.Conv1d(self.feature_dim, self.feature_dim, kernel_size=1)
+        self.proj = nn.ModuleList([deepcopy(self.merge) for _ in range(3)])
+
+    def forward(
+        self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor, output_attentions: bool = False
+    ) -> Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor]]:
+        batch_dim = query.size(0)
+        query, key, value = [
+            layer(x).view(batch_dim, self.dim, self.num_heads, -1) for layer, x in zip(self.proj, (query, key, value))
+        ]
+        x, attention_probs = attention(query, key, value)
+        output = self.merge(x.contiguous().view(batch_dim, self.dim * self.num_heads, -1))
+
+        output = (output, attention_probs) if output_attentions else (output,)
+
+        return output
+
+
+class SuperGlueAttentionalPropagation(nn.Module):
+    def __init__(
+        self,
+        config: SuperGlueConfig,
+    ):
+        super().__init__()
+        self.descriptor_dim = config.descriptor_dim
+        self.num_heads = config.num_heads
+        self.attention = SuperGlueMultiHeadAttention(config)
+        self.mlp = SuperGlueMultiLayerPerceptron(
+            config,
+            [self.descriptor_dim * 2, self.descriptor_dim * 2, self.descriptor_dim],
+        )
+        nn.init.constant_(self.mlp.layers[-1].bias, 0.0)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        source: torch.Tensor,
+        output_attentions: bool = False,
+        output_hidden_states: bool = False,
+    ) -> tuple:
+        attention_outputs = self.attention(x, source, source, output_attentions=output_attentions)
+        output = attention_outputs[0]
+        attention = attention_outputs[1:]
+
+        output = torch.cat([x, output], dim=1)
+        layer_outputs = self.mlp(output, output_hidden_states=output_hidden_states)
+
+        last_hidden_state = layer_outputs[0]
+        hidden_states = layer_outputs[1]
+
+        return last_hidden_state, hidden_states, attention
+
+
+class SuperGlueAttentionalGNN(nn.Module):
+    def __init__(
+        self,
+        config: SuperGlueConfig,
+    ):
+        super().__init__()
+        self.descriptor_dim = config.descriptor_dim
+        self.num_heads = config.num_heads
+        self.layers_types = config.gnn_layers_types
+        self.num_layers = len(self.layers_types)
```

Comment:
no need to set these attributes 

```suggestion
        self.layers_types = config.gnn_layers_types
```

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/29886/commits/601da1cf0756cef7339d4dd33faeefbfcd073ac9)